File: languages/csharp/AppEncryption/Crypto/Keys/CryptoKey.cs
using System;

namespace GoDaddy.Asherah.Crypto.Keys
{
    public abstract class CryptoKey : IDisposable
    {
        public abstract DateTimeOffset GetCreated();

        public abstract void WithKey(Action<byte[]> actionWithKey);

        public abstract TResult WithKey<TResult>(Func<byte[], TResult> actionWithKey);

        public abstract void Dispose();

        public abstract bool IsRevoked();

        public abstract void MarkRevoked();
    }
}

File: languages/java/app-encryption/src/main/java/com/godaddy/asherah/crypto/keys/CryptoKey.java
package com.godaddy.asherah.crypto.keys;

import java.time.Instant;
import java.util.function.Consumer;
import java.util.function.Function;

import com.godaddy.asherah.appencryption.utils.SafeAutoCloseable;

public abstract class CryptoKey implements SafeAutoCloseable {
  public abstract Instant getCreated();

  public abstract boolean isRevoked();

  public abstract void markRevoked();

  public abstract void withKey(Consumer<byte[]> action);

  public abstract <T> T withKey(Function<byte[], T> action);
}
